File: node-api/src/utils/groqClient.js
const Groq = require('groq-sdk');
const env = require('../config/env');

// Matches the model python-service's ai.py/resume.py/interviews.py/assessments.py
// all use — kept as one constant so a future model bump happens in one place.
const MODEL = 'llama-3.3-70b-versatile';

let client = null;

function isGroqConfigured() {
  return Boolean(env.groq.apiKey);
}

function getClient() {
  if (!client) {
    if (!isGroqConfigured()) {
      throw new Error('GROQ_API_KEY is not configured');
    }
    client = new Groq({ apiKey: env.groq.apiKey });
  }
  return client;
}

// Callers decide how to react to a missing key (check isGroqConfigured() first) —
// python-service does this two different ways per endpoint (soft passthrough
// fallback in ai.py/interviews.py, hard 500 in resume.py), so that choice isn't
// baked into this wrapper.
async function groqComplete({ systemPrompt, userPrompt, temperature = 0.7, maxTokens = 1024, jsonResponse = false }) {
  const groq = getClient();
  const completion = await groq.chat.completions.create({
    model: MODEL,
    temperature,
    max_tokens: maxTokens,
    messages: [
      { role: 'system', content: systemPrompt },
      { role: 'user', content: userPrompt },
    ],
    ...(jsonResponse ? { response_format: { type: 'json_object' } } : {}),
  });
  const content = completion.choices[0]?.message?.content || '';
  return jsonResponse ? JSON.parse(content) : content;
}

module.exports = { isGroqConfigured, groqComplete, MODEL };

File: ai-service/app/groq_client.py
import json
import logging

import httpx

from app.config import get_settings

logger = logging.getLogger("ai-service.groq")


class GroqError(Exception):
    """Raised when a configured Groq call fails outright (network error,
    non-2xx response, or unparseable JSON when jsonResponse was requested).
    Callers translate this into a 502 — distinct from "not configured",
    which routes check for themselves via settings.is_groq_configured
    before ever reaching this module."""


async def groq_complete(
    system_prompt: str,
    user_prompt: str,
    *,
    temperature: float = 0.7,
    max_tokens: int = 1024,
    json_response: bool = False,
) -> str | dict:
    settings = get_settings()
    body = {
        "model": settings.groq_model,
        "temperature": temperature,
        "max_tokens": max_tokens,
        "messages": [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt},
        ],
    }
    if json_response:
        body["response_format"] = {"type": "json_object"}

    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            resp = await client.post(
                f"{settings.groq_base_url}/chat/completions",
                headers={"Authorization": f"Bearer {settings.groq_api_key}"},
                json=body,
            )
            resp.raise_for_status()
            data = resp.json()
    except httpx.HTTPError as exc:
        logger.error("Groq request failed: %s", exc)
        raise GroqError(str(exc)) from exc

    content = data["choices"][0]["message"]["content"] or ""
    if not json_response:
        return content
    try:
        return json.loads(content)
    except json.JSONDecodeError as exc:
        logger.error("Groq returned non-JSON content for a json_response request: %s", exc)
        raise GroqError(f"Groq returned invalid JSON: {exc}") from exc
